feat: wip randomx miner

brianp opened the pull request:
Description
---
Provide a randomx miner.

Motivation and Context
---
Don't have it flagged as malware when shipping.

How Has This Been Tested?
---
NA

What process can a PR reviewer use to test or verify this change?
---

<!-- Checklist -->
<!-- 1. Is the title of your PR in the form that would make nice release notes? The title, excluding the conventional commit
tag, will be included exactly as is in the CHANGELOG, so please think about it carefully. -->


Breaking Changes
---

- [x] None
- [ ] Requires data directory on base node to be deleted
- [ ] Requires hard fork
- [ ] Other - Please specify

<!-- Does this include a breaking change? If so, include this line as a footer -->
<!-- BREAKING CHANGE: Description what the user should do, e.g. delete a database, resync the chain -->

brianp committed:
Add scaffold for randomx_miner

brianp committed:
Communicate with monero node over proxy

brianp committed:
refactor requests to the proxy

brianp committed:
Calculate the hash

brianp committed:
increment the blockhashing blog nonce

brianp committed:
Error refactoring

brianp committed:
Allow the RandomX factory to accept params

Specifically the cache and dataset needed for max mode mining.

brianp committed:
Add to CI

brianp committed:
submit the block with successful hash

brianp committed:
Update the mined block timestamp

brianp committed:
Initialize empty caches on lite mode

brianp committed:
Add cli options for node and wallet address

brianp committed:
Update config options

brianp committed:
Enhance the config options with more usability

brianp committed:
Re-write for better thread management

brianp committed:
Add support for shared caches when multithreading

brianp committed:
Better bail out for new templates

brianp committed:
Enhance logging and include thread index

brianp committed:
Remove unused includes

brianp committed:
formatting

brianp committed:
machete

brianp committed:
lints

SWvheerden commented on applications/randomx_miner/Cargo.toml:
```diff
@@ -0,0 +1,37 @@
+description = "The Tari merge mining proxy for xmrig"
+repository = "https://github.com/tari-project/tari"
+license = "BSD-3-Clause"
+version = "1.3.0-pre.0"
```
```suggestion
version = "1.4.0-pre.0"
```

brianp committed:
tari deps version bump

SWvheerden commented on applications/randomx_miner/src/run_miner.rs:
```diff
@@ -0,0 +1,253 @@
+
+    let timestamp_position = 8;
+    let timestamp_bytes: [u8; 4] = (EpochTime::now().as_u64() as u32).to_le_bytes();
+    blockhashing_bytes[timestamp_position..timestamp_position + 4].copy_from_slice(&timestamp_bytes);
```
I would not update the timestamp every cycle, its wasting time

SWvheerden commented on applications/randomx_miner/src/run_miner.rs:
```diff
@@ -0,0 +1,253 @@
+                    thread_number, config.template_refresh_interval_ms
+                );
+                let new_block_template =
+                    runtime.block_on(get_block_template(client, node_address, monero_wallet_address))?;
```
I would not do this here, I would put this on a separate loop, and let that loop check this. If the template changes then we send an interrupt that the mining threads, having each thread check this wastes mining cycles

SWvheerden commented on applications/randomx_miner/src/run_miner.rs:
```diff
@@ -0,0 +1,253 @@
+                    break;
+                }
+            }
+            let (difficulty, hash) = mining_cycle(blockhashing_bytes.clone(), nonce as u32, vm.clone())?;
```
let it mut burrow rather than clone

SWvheerden commented on applications/randomx_miner/src/run_miner.rs:
```diff
@@ -0,0 +1,253 @@
+                }
+            }
+            let (difficulty, hash) = mining_cycle(blockhashing_bytes.clone(), nonce as u32, vm.clone())?;
+            stats_store.inc_hashed_count();
```
this will be bad as well, I think we need to cache the results only update every 1000 hashes or so. Or some interval. Not every hash. Every thread will want to update every hash

brianp committed:
Less noisy console

brianp committed:
Fix refresh limit

brianp committed:
Use thread::scope to avoid cloning

brianp committed:
Use unique keys when creating vms

But use shared keys for caches and datasets. We use the key for local
caching of the vms but this key is also passed to the cache naming. We
need our cache to be unique but caches to be the shared.

brianp committed:
Only report block found on success submit

SWvheerden commented on applications/randomx_miner/src/run_miner.rs:
```diff
@@ -0,0 +1,270 @@
+    let nonce_position = 38;
+    blockhashing_bytes[nonce_position..nonce_position + 4].copy_from_slice(&nonce.to_le_bytes());
+
+    let timestamp_position = 8;
+    let timestamp_bytes: [u8; 4] = u32::try_from(EpochTime::now().as_u64())?.to_le_bytes();
+    blockhashing_bytes[timestamp_position..timestamp_position + 4].copy_from_slice(&timestamp_bytes);
```
I would not do this every hash, it wastes CPU cycles

brianp committed:
Ensure the right loops are broken

Make sure the seed hash is checked and updated when needed. Ensure the
block template is updated when needed.

brianp committed:
Move blocking code out of the mining cycle

- Moves the stats incrementer to cache a number of counts and write
  later instead of on each cycle
- Starts a thread to watch the block template and notify the thread when
  it changes

brianp committed:
Remove the randomx miner application

brianp committed:
Remove remnamts

brianp commented:
Closing this PR in favour of: https://github.com/tari-project/tari/pull/6537/files

and a new repo: https://github.com/tari-project/clythor
